flink-connectors/flink-connector-gcp-pubsub/src/main/java/org/apache/flink/streaming/connectors/gcp/pubsub/PubSubSource.java

```diff
@@ -0,0 +1,296 @@
+	protected final DeserializationSchema<OUT> deserializationSchema;
+	protected final PubSubSubscriberFactory pubSubSubscriberFactory;
+	protected final Credentials credentials;
+	protected final String projectSubscriptionName;
+	protected final int maxMessagesPerPull;
+
+	protected transient boolean deduplicateMessages;
+	protected transient SubscriberStub subscriber;
+	protected transient PullRequest pullRequest;
+	protected transient EventLoopGroup eventLoopGroup;
+
+	protected transient volatile boolean isRunning;
+	protected transient volatile ApiFuture<PullResponse> messagesFuture;
+
+	PubSubSource(DeserializationSchema<OUT> deserializationSchema, PubSubSubscriberFactory pubSubSubscriberFactory, Credentials credentials, String projectSubscriptionName, int maxMessagesPerPull) {
+		super(String.class);
+		this.deserializationSchema = deserializationSchema;
+		this.pubSubSubscriberFactory = pubSubSubscriberFactory;
+		this.credentials = credentials;
+		this.projectSubscriptionName = projectSubscriptionName;
+		this.maxMessagesPerPull = maxMessagesPerPull;
+	}
+
+	@Override
+	public void open(Configuration configuration) throws Exception {
+		super.open(configuration);
+		if (hasNoCheckpointingEnabled(getRuntimeContext())) {
+			throw new IllegalArgumentException("The PubSubSource REQUIRES Checkpointing to be enabled and " +
+				"the checkpointing frequency must be MUCH lower than the PubSub timeout for it to retry a message.");
+		}
+
+		getRuntimeContext().getMetricGroup().gauge("PubSubMessagesProcessedNotAcked", this::getOutstandingMessagesToAck);
+
+		this.eventLoopGroup = new NioEventLoopGroup();
+		this.subscriber = pubSubSubscriberFactory.getSubscriber(eventLoopGroup, credentials);
+		this.deduplicateMessages = getRuntimeContext().getNumberOfParallelSubtasks() == 1;
+		this.isRunning = true;
+		this.pullRequest = PullRequest.newBuilder()
+										.setMaxMessages(100)
+										.setReturnImmediately(false)
+										.setSubscription(projectSubscriptionName)
+										.build();
+	}
+
+	private boolean hasNoCheckpointingEnabled(RuntimeContext runtimeContext) {
+		return !(runtimeContext instanceof StreamingRuntimeContext && ((StreamingRuntimeContext) runtimeContext).isCheckpointingEnabled());
+	}
+
+	@Override
+	protected void acknowledgeSessionIDs(List<String> acknowledgementIds) {
+		if (acknowledgementIds.isEmpty() || !isRunning) {
+			return;
+		}
+
+		AcknowledgeRequest acknowledgeRequest =
+			AcknowledgeRequest.newBuilder()
+								.setSubscription(projectSubscriptionName)
+								.addAllAckIds(acknowledgementIds)
+								.build();
+		subscriber.acknowledgeCallable().call(acknowledgeRequest);
+	}
+
+	@Override
+	public void run(SourceContext<OUT> sourceContext) throws Exception {
+		while (isRunning) {
+			messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
+			try {
+				processMessage(sourceContext, messagesFuture.get().getReceivedMessagesList());
```

Comment:
If the message process takes long to finish, the next pull request will also be delayed. Performance wise, it might be better to have something like below:
```
messagesFuture = subscriber.pullCallable().futureCall(pullRequest);
while (isRunning) {
    try {
        PullResponse response = messageFuture.get();
        // start the next pull while processing the current response.
        messageFuture = subscriber.pullCallable().futureCall(pullRequest);
	processMessage(sourceContext, response.getReceivedMessagesList());
    } catch (InterruptedException | CancellationException e) {
	awaitSubscriberTermination();
	return;
    }
}
```

Reply:
Interesting idea, nice catch! I wanted to see how much this would improve performance and I didn't improve it all that much to my suprise.

What I did notice that the amount of messages returned per pull was (on average) larger - which would mean less requests / pulls will be done so that's good.

I've changed it, thanks! 👍

Reply:
Thanks for the update.

The throughput gain will depend on two things:

1. How long it takes to process the messages. If the `processMessage` runs quickly, then there won't be much throughput improvement.

2. How long it takes to fetch the messages. If the message fetching takes very small amount of time, there won't be much throughput improvement either.